polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -462,211 +572,204 @@ impl FragmentNode {
 		&self.scope
 	}
 
-	/// Returns the number of candidates in the chain
-	pub(crate) fn len(&self) -> usize {
-		self.candidates.len()
+	/// Returns the number of candidates in the best backable chain.
+	pub fn best_chain_len(&self) -> usize {
+		self.best_chain.chain.len()
+	}
+
+	/// Returns the number of candidates in unconnected potential storage.
+	pub fn unconnected_len(&self) -> usize {
+		self.unconnected.len()
 	}
 
-	/// Whether the candidate exists.
-	pub(crate) fn contains_candidate(&self, candidate: &CandidateHash) -> bool {
-		self.candidates.contains(candidate)
+	/// Whether the candidate exists as part of the unconnected potential candidates.
+	pub fn contains_unconnected_candidate(&self, candidate: &CandidateHash) -> bool {
+		self.unconnected.contains(candidate)
 	}
 
 	/// Return a vector of the chain's candidate hashes, in-order.
-	pub(crate) fn to_vec(&self) -> Vec<CandidateHash> {
-		self.chain.iter().map(|candidate| candidate.candidate_hash).collect()
+	pub fn best_chain_vec(&self) -> Vec<CandidateHash> {
+		self.best_chain.chain.iter().map(|candidate| candidate.candidate_hash).collect()
 	}
 
-	/// Try accumulating more candidates onto the chain.
-	///
-	/// Candidates can only be added if they build on the already existing chain.
-	pub(crate) fn extend_from_storage(&mut self, storage: &CandidateStorage) {
-		self.populate_chain(storage);
+	/// Return a vector of the unconnected potential candidate hashes, in arbitrary order.
+	pub fn unconnected(&self) -> impl Iterator<Item = &CandidateEntry> {
+		self.unconnected.candidates()
 	}
 
-	/// Returns the hypothetical state of a candidate with the given hash and parent head data
-	/// in regards to the existing chain.
-	///
-	/// Returns true if either:
-	/// - the candidate is already present
-	/// - the candidate can be added to the chain
-	/// - the candidate could potentially be added to the chain in the future (its ancestors are
-	///   still unknown but it doesn't violate other rules).
-	///
-	/// If this returns false, the candidate could never be added to the current chain (not now, not
-	/// ever)
-	pub(crate) fn hypothetical_membership(
-		&self,
-		candidate: HypotheticalCandidate,
-		candidate_storage: &CandidateStorage,
-	) -> bool {
-		let candidate_hash = candidate.candidate_hash();
+	/// Return whether this candidate is backed in this chain or the unconnected storage.
+	pub fn is_candidate_backed(&self, hash: &CandidateHash) -> bool {
+		self.best_chain.candidates.contains(hash) ||
+			matches!(
+				self.unconnected.by_candidate_hash.get(hash),
+				Some(candidate) if candidate.state == CandidateState::Backed
+			)
+	}
 
-		// If we've already used this candidate in the chain
-		if self.candidates.contains(&candidate_hash) {
-			return true
+	/// Return a new [`CandidateStorage`] containing all the candidates from this `FragmentChain`,
+	/// as well as the unconnected ones.
+	pub fn as_candidate_storage(&self) -> CandidateStorage {
```

Comment:
I replaced one use of `as_candidate_storage` (when marking a candidate as backed) with your suggestion.
However, the other place where we call this can't consume it (during active leaf update), because subsequent forks need to have access to the parent's candidates. I'll see how I can make this prettier (maybe by using your advance_scope suggestion)